### PR TITLE
Init config file structure for recipes

### PR DESCRIPTION
**Pull request by natolambert**
Here, I collected some outputs from the original argparse and the new config with the new script `sh scripts/finetune_with_accelerate_1gpu.sh`. Then, for each N gpus training script, we can keep recipes in `train_configs/sft` or `train_configs/dpo`.

The outputs are slightly different because new defaults (which are fine):

Original:
```
add_bos: False
checkpointing_steps: None
clip_grad_norm: -1
config_name: None
dataset_config_name: None
dataset_name: None
gradient_accumulation_steps: 4
gradient_checkpointing: False
learning_rate: 2e-05
logging_steps: 1
lora_alpha: 16
lora_dropout: 0.1
lora_rank: 64
low_cpu_mem_usage: False
lr_scheduler_type: linear
max_seq_length: 4096
max_train_steps: None
model_name_or_path: allenai/OLMo-1.7-1B-hf
model_revision: main
num_train_epochs: 2
output_dir: output/tulu_v2_1B/
overwrite_cache: False
per_device_train_batch_size: 32
preprocessing_num_workers: 128
reduce_loss: mean
report_to: tensorboard
resume_from_checkpoint: None
seed: 42
timeout: 1800
tokenizer_name: allenai/OLMo-1.7-1B-hf
tokenizer_revision: main
train_file: data/processed/tulu_v2/tulu_v2_data.jsonl
trust_remote_code: False
use_8bit_optimizer: False
use_flash_attn: True
use_lora: False
use_qlora: False
use_slow_tokenizer: True
wandb_entity: None
warmup_ratio: 0.03
weight_decay: 0.0
with_tracking: True
```

New:
```
add_bos: False
checkpointing_steps: None
clip_grad_norm: -1
config_name: None
dataset_config_name: None
dataset_name: None
gradient_accumulation_steps: 4
gradient_checkpointing: False
learning_rate: 2e-05
logging_steps: 1
lora_alpha: 16
lora_dropout: 0.1
lora_rank: 64
low_cpu_mem_usage: False
lr_scheduler_type: linear
max_seq_length: 4096
max_train_steps: None
model_name_or_path: allenai/OLMo-1.7-1B-hf
model_revision: main
num_train_epochs: 2
output_dir: output/tulu_v2_1B/
overwrite_cache: False
per_device_train_batch_size: 32
preprocessing_num_workers: 128
reduce_loss: mean
report_to: tensorboard
resume_from_checkpoint: None
seed: 42
timeout: 1800
tokenizer_name: allenai/OLMo-1.7-1B-hf
tokenizer_revision: None
train_file: data/processed/tulu_v2/tulu_v2_data.jsonl
trust_remote_code: False
use_8bit_optimizer: False
use_flash_attn: True
use_lora: False
use_qlora: False
use_slow_tokenizer: True
wandb_entity: None
warmup_ratio: 0.03
weight_decay: 0.0
with_tracking: True
```

### CODE REVIEW

File: open_instruct/finetune.py [hunk 1]
```diff
@@ -400,9 +159,260 @@ def save_with_accelerate(accelerator, model, tokenizer, output_dir, args):
             safe_serialization=False
         )
 
+# def parse_args():
```
Comment (hamishivi) [inline open_instruct/finetune.py]: nit: probably good to remove big commented blocks before merging.

File: open_instruct/utils.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    # )
+    use_slow_tokenizer: bool = field(
+        default=True,
+        metadata={"help": "Whether to use one of the slow tokenizer or not (which is then fast tokenizer)."},
+    )
+    model_revision: str = field(
+        default="main",
+        metadata={"help": "The specific model version to use (can be a branch name, tag name or commit id)."},
+    )
+    # token: str = field(
```
Comment (hamishivi) [inline open_instruct/utils.py]: why removing this?

File: open_instruct/utils.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        },
+    )
+    dataset_name: Optional[str] = field(
+        default=None, metadata={"help": "The name of the dataset to use (via the datasets library)."}
+    )
+    dataset_config_name: Optional[str] = field(
+        default=None, metadata={"help": "The configuration name of the dataset to use (via the datasets library)."}
+    )
+    train_file: Optional[str] = field(default=None, metadata={"help": "The input training data file (a json/jsonl file)."})
+    # max_train_samples: Optional[int] = field(
```
Comment (hamishivi) [inline open_instruct/utils.py]: why remove this?

File: train_configs/sft/default.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+model_name_or_path: allenai/OLMo-1.7-1B-hf
+model_revision: main
+use_flash_attn: true
+tokenizer_name: allenai/OLMo-1.7-1B-hf
+use_slow_tokenizer: true
+train_file: data/processed/tulu_v2/tulu_v2_data.jsonl
```
Comment (hamishivi) [inline train_configs/sft/default.yaml]: could we just swap this to point to the hf uploaded version? makes it even more portable :)

File: train_configs/sft/tulu2_70b.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+lr_scheduler_type: linear
+warmup_ratio: 0.03
+weight_decay: 0.0
+num_train_epochs: 3
+output_dir: output/tulu_v2_70b/
+with_tracking: true
+report_to:
+  - wandb
+logging_steps: 1
+checkpointing_steps: epoch
```
Comment (hamishivi) [inline train_configs/sft/tulu2_70b.yaml:23]: probably also need gradient checkpointing?

File: scripts/finetune_with_accelerate_new.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+# You can also set --gradient_checkpointing or use `stage3_offloading_accelerate.conf` to save memory, 
+# but it will trade off speed.
+accelerate launch \
+    --mixed_precision bf16 \
+    --num_machines 1 \
+    --num_processes $NUM_GPUS \
+    --use_deepspeed \
+    --deepspeed_config_file ds_configs/stage3_no_offloading_accelerate.conf \
+    open_instruct/finetune.py \
+    "$1"
```
Comment (hamishivi) [inline scripts/finetune_with_accelerate_new.sh]: this should be 2?

File: train_configs/sft/olmo_7b_17_sft.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+model_name_or_path: allenai/OLMo-1.7-7B-hf
+model_revision: main
+use_flash_attn: true
+tokenizer_name: allenai/OLMo-1.7-7B-hf
+use_slow_tokenizer: true
```
Comment (hamishivi) [inline train_configs/sft/olmo_7b_17_sft.yaml]: this should be false for olmo models.

File: open_instruct/utils.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    )
+    lora_alpha: float = field(
+        default=16,
+        metadata={"help": "The alpha parameter of lora."},
+    )
+    lora_dropout: float = field(
+        default=0.1,
+        metadata={"help": "The dropout rate of lora modules."},
+    )
+    lr_scheduler_type: str = field(
```
Comment (hamishivi) [inline open_instruct/utils.py:145]: Before we had this as a `SchedulerType` arg, not a string, and so the code errors out due to this being a string... unsure of the quickest way to fix.

**Review by hamishivi — COMMENTED** [review]
I think this still needs a bit more polish, but I think the structure seems good!

File: open_instruct/finetune.py (hunk 1 above)
Comment (natolambert) [reply open_instruct/finetune.py]: Yeah should've made this clear -- left in for testing / making it easier to compare code, but removing pre merge.

File: open_instruct/utils.py (hunk 2 above)
Comment (natolambert) [reply open_instruct/utils.py]: Commented args here were in `finetune_trainer.py` and not `finetune.py`. Can go either way, but wasn't used in `finetune.py`, so this was an unclear way of cleaning.

File: open_instruct/utils.py (hunk 3 above)
Comment (natolambert) [reply open_instruct/utils.py]: Similar to above, not in `finetune.py`. Can add it in.

File: train_configs/sft/default.yaml (hunk 4 above)
Comment (natolambert) [reply train_configs/sft/default.yaml]: Yes fixing this, was a typo / example for debugging not actual training.

File: scripts/finetune_with_accelerate_new.sh (hunk 6 above)
Comment (natolambert) [reply scripts/finetune_with_accelerate_new.sh]: Yup, miscopied.

File: open_instruct/utils.py (hunk 8 above)
Comment (natolambert) [reply open_instruct/utils.py:145]: I noticed this, but looking at the docs it made it so string was valid. Double checking.

Comment (natolambert) [reply open_instruct/utils.py:145]: https://huggingface.co/docs/transformers/en/main_classes/optimizer_schedules#transformers.get_scheduler

`name (str or SchedulerType)`

File: open_instruct/utils.py (hunk 2 above)
Comment (hamishivi) [reply open_instruct/utils.py]: Ah, right. I think rn for gated models we just pick up on HF_TOKEN env var, so can remove this.

File: open_instruct/utils.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+        default=False,
+        metadata={
+            "help": (
+                "Whether or not to allow for custom models defined on the Hub in their own modeling files. This option"
+                "should only be set to `True` for repositories you trust and in which you have read the code, as it will "
+                "execute code present on the Hub on your local machine."
+            )
+        },
+    )
+    # torch_dtype: Optional[str] = field(
```
Comment (hamishivi) [inline open_instruct/utils.py]: similarly can just remove.

File: open_instruct/utils.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+    max_train_samples: Optional[int] = field(
+        default=None,
+        metadata={
+            "help": (
+                "For debugging purposes or quicker training, truncate the number of training examples to this "
+                "value if set."
+            )
+        },
+    )
+    # streaming: bool = field(default=False, metadata={"help": "Enable streaming mode"})
```
Comment (hamishivi) [inline open_instruct/utils.py]: also can just remove.

File: open_instruct/utils.py (hunk 8 above)
Comment (hamishivi) [reply open_instruct/utils.py:145]: cool! seems to work in  my testing now.

File: scripts/finetune_with_accelerate_new.sh [hunk 11]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Comment (hamishivi) [inline scripts/finetune_with_accelerate_new.sh]: nit: `_new` isn't very informative, maybe use `_config`? or something?

**Review by hamishivi — APPROVED** [review]
Seems good! Gave it a little test and worked okay.
